### ISSUE TITLE
Problem with the new autoRowsToHeight property

### ISSUE THREAD

**Issue by Josua2012** [maxim-saplin/data_table_2#29]
Marvelous. I love the new autoRowsToHeight property. :)

However, there is a serious problem that needs to be corrected, since in my case it causes me to not be able to use the new autoRowsToHeight property.
Changing the number of rows does not raise the onRowsPerPageChanged event.
I use it to fetch records from a nodejs server.
It would also be very interesting if the execution of onRowsPerPageChanged could be debounced (RowsPerPageChangedDebounced??), that is, that the onRowsPerPageChanged event does not execute for each change of rowsPerPage in the case of several property changes very quickly when changing the height of the PaginatedDataTable2 widget.
Check out the following video: https://recordit.co/JT0UWm2y6x

If you have any questions about what I have told you, tell me.

Thank you very much for your hard work. 

Regards,
Jose

**Comment by maxim-saplin** [comment]
Indeed missed that, thanks for pointing out. Will look into that in the evening, shouldn't be hard.
Regarding debouncing, I wouldn't put that into widget (you don't find often debounce capabilities in message handlers don't you ;) - you better solve this via a separate capability, here's an example of debounce mixin I used in one of my projects: https://github.com/maxim-saplin/dikt/blob/cdb361a258c0c68a11595b3ffb9ead5dd4ce5366/lib/common/debounceMixin.dart

**Comment by Josua2012** [comment]
Thank you very much for the suggestion. :)

**Comment by maxim-saplin** [comment]
I've just commited the fix to Master. Could you please check via referencing github sources directly?
  data_table_2:
    git: https://github.com/maxim-saplin/data_table_2

**Comment by Josua2012** [comment]
I have done some tests and now it works correctly, however, it would be fine if autoRowsToHeight is true and rowsperpage parameter has also been used, this property (rowsperpage) will be ignored in the code, since the simultaneous use of the 2 properties gives problems:
```
Assertion failed: 
paginted_data_table_2.dart: 73:20
availableRowsPerPage.contains(rowsPerPage)
is not true
```
In my case I also have some refresh problems (flicker) as I am trying to simulate the async paged loading of rows from the server, but I guess this will work fine when you make the "async-table" update of your repository available. 

Thank you very much for everything.
They are very important innovations in their package and all the developers will be very grateful to you. 💯 🥇

**Comment by maxim-saplin** [comment]
Could you please tell more about the assertion error and exact steps to see it.

Regarding flicker, I'd advice you testing for performance in Release mode and in Chrome (if in the Web). Flutter's Tables are slow in Debug, also Flutter Web is known to have UI junk issues in browsers other than Chrome.

**Comment by Josua2012** [comment]
The following code would cause the error that I have mentioned:
                 rowsPerPage: powerArduinoProvider.pagPowerArduino.rowsPerPage,
                 onRowsPerPageChanged: (value) {
                        EasyDebounce.debounce(
                            'rowsperpagechanged-debouncer',
                            Duration(milliseconds: 300),
                            () =>
                                powerArduinoProvider.adjustRowsPerPage(value!));
                      },

Removing the line of code "rowsPerPage" prevents the error from occurring.

I am using Flutter web in chrome, but the problem is my fault because of the system that I had to do to simulate the paginated loading of rows from the server, it is not an intrinsic problem of Flutter tables, that is why I was telling you that your implementation for asynchronous loading of rows from the server should solve this problem. 

Thanks.

**Comment by maxim-saplin** [comment]
Will look into the assertion, it is not good there's one with auto rows turned on. Though it doesn't make sense to have `autoRowsToHeight` always set to true and at the same time having `rowsPerPage` set to some value. Does it?

Async is in progress, hope to have it done over the weekend.

**Comment by Josua2012** [comment]
Of course, autoRowsToHeight and rowsPerPage should not be used simultaneously, but if a developer is testing their application they may go a bit crazy to find out what is happening if they are using both properties at the moment. IMHO, if autoRowsToHeight is true the rowsPerPage property should be ignored internally in your code.
If you could have the asynchronous rows loading version by next week that would be great. 👍 
Many thanks. 
Regards,
Jose

**Comment by maxim-saplin** [comment]
The assertion is fixed on master, you can check it

**Comment by Josua2012** [comment]
Now it no longer produces the error. 👍 
But I have discovered another problem, the initialFirstRowIndex property does not seem to work correctly if autoRowsToHeight is true.
When initially viewing the table, I want to go to the last page, for this I do the following: 
```
                initialFirstRowIndex:
                    logProvider.pagLog.getInitialFirstRowIndex(),
```
This works correctly if autoRowsToHeight is false. 

Can you think of something why this may not work correctly? 

Thanks.

**Comment by Josua2012** [comment]
Another problem, the onRowsPerPageChanged event does not initially fire when the PaginatedDataTable2 is created, so it is impossible to initially know the rows that the table has. This is very useful if autoRowsToHeight is true, since if it is false the PaginatedDataTable component is always initialized with a default number of rows.

**Comment by Josua2012** [comment]
The `initialFirstRowIndex `property also doesn't work correctly for me with `autoRowsToHeight to true` if I use the mouse wheel to navigate between pages. 
The code is the following:
```
 child: Listener(
        onPointerSignal: (pointerSignal) {
          if (pointerSignal is PointerScrollEvent) {
            // do something when scrolled
            //PointerScrollEvent pp = pointerSignal as PointerScrollEvent;

            //print('Scrolled ${pp.scrollDelta.dy}');

            if (pointerSignal.scrollDelta.dy > 0) {
              usersProvider.nextPage();
            }
            if (pointerSignal.scrollDelta.dy < 0) {
              usersProvider.priorPage();
            }
          }
        },
        child: PaginatedDataTable2(
          key: usersProvider.formKey,
          autoRowsToHeight: true,
          initialFirstRowIndex: usersProvider.indexFirstRowPage,
          columnSpacing: 5,
 onRowsPerPageChanged: (value) {
            usersProvider.rowsPerPage = value!;
          },
          onPageChanged: (page) {
            usersProvider.indexFirstRowPage = page;
          },

// provider

Future nextPage() async {
    final newIndexFirstRowPage =
        indexFirstRowPage + formKey.currentState!.effectiveRowsPerPage;
    if (newIndexFirstRowPage <= totalRows) {
      indexFirstRowPage = newIndexFirstRowPage;
      notifyListeners();
    }
  }

  Future priorPage() async {
    int newIndexFirstRowPage =
        indexFirstRowPage - formKey.currentState!.effectiveRowsPerPage;
    if ((newIndexFirstRowPage < 0) && (indexFirstRowPage > 0)) {
      newIndexFirstRowPage = 0;
    }
    if (newIndexFirstRowPage >= 0) {
      indexFirstRowPage = newIndexFirstRowPage;
      notifyListeners();
    }
  }
```

The provider's `nextPage` and `priorPage` methods simply calculate the row index of the next or previous page to assign to the `initialFirstRowIndex `property.
As you will see in the video (https://recordit.co/VW6fWRfN5z), everything occurs correctly, however navigation between the pages does not occur.
Strangely, if I don't use the mouse wheel and assign a constant value to the `initialFirstRowIndex` property if the page that has that row at the top is displayed.
I'm going a little crazy with all of this. 

Thanks for your help.

**Comment by maxim-saplin** [comment]
onRowsPerPageChanged is now called everytime. initialFirstRowIndex seems to be working just fine, please check the examples

**Comment by Josua2012** [comment]
As I was saying and you can see this problem in the video that I sent you, when Listener is used to detect the mouse wheel, for some reason initialFirstRowIndex does not work correctly. If Listener is not used to detect the mouse wheel then it works fine.

**Comment by maxim-saplin** [comment]
Unfortunately I don't have time to dive into you code snippets and try recreating some repro steps and debuggable sample. I still don't see any isolated issue which demonstrates how certain feature of the widget doesn't work as intended so I could fix it.

You just thrown some code of your own my way and say 'go figure' :)

**Comment by Josua2012** [comment]
Hi Maxim,
I have used your example code and recreated the problem.
[example.zip](https://github.com/maxim-saplin/data_table_2/files/6905892/example.zip)
You can also watch the following video:  https://recordit.co/ST1tNtUNGC
If you have any questions, tell me.
Thanks for the help.

**Comment by maxim-saplin** [comment]
Great, will look into it later

**Comment by Josua2012** [comment]
Perfect. Thanks Maxim.

**Comment by maxim-saplin** [comment]
It seems to me that using the `initialFirstRowIndex` for switching pages is not the best solution, the 'initial' part suggests that the widget use the value only once... Not sure why it worked before and now doesn't work. 

It seems to me more reasonble to create a propper API to let control widgets paginator externally rather than try to fix the workaround thay is now broken. Some time ago that was mentioned in a separate issue regarding switching pages via scroller.

The idea is that there will be a class PaginatorController whichj you will be able to create and set to widget's property (just like you do with other controllers like ScrollController). The controller will have the methods providing the following capabilities:
- Go to Next page
- Go to Previous Page
- Get current page index
- Go to page Index
- Get page size
- Set page size to X (if no auto rows option is enabled)

This would be a general solutuion that will let you implement your feature with no hassle. Will try to put smth together sonner

**Comment by Josua2012** [comment]
Yes. This would be perfect. :)
You could even put the following function:
Go to the page that contains the index row
or
Get the index of the page that contains the index row

Thanks.

**Comment by maxim-saplin** [comment]
https://github.com/maxim-saplin/data_table_2/tree/page-controller - I've created the controller in 'page-controller ' branch which gives the propper way to control widget. Could you please give it a try? There're also an example which demostrates it's capabilities via a custom paginator.

**Comment by maxim-saplin** [comment]
JIC, now the feature is available on master

**Comment by Josua2012** [comment]
Maxim,
It works perfectly. 💯 
Many thanks.
Now it would only be missing the async rows paginated functionality from the server to be great. :)
Again, thank you very much. 👍 🥇

**Comment by maxim-saplin** [comment]
Great, 2.1.1 with fixes is on pub.dev
